README.Rmd

```diff
@@ -49,20 +49,53 @@ This identifier leverages the familiar URI format of other identifiers, indicati
 register(co2_url)
 ```
 
 Note that upon registering the URL, hash-archive.org has returned the very same identifier, confirming this is the same content as our local copy. Such an archive is nothing more than a simple key-value store, indicating which URL(s) contain our content.  In this way, it can act much like a DOI re-direct, turing our identifier into resolvable, downloadable URL.  Unlike the DOI system though, this system is completely distributed -- we need not rely on any single authority to issue the identifier, and we need not resolve the identifier to a single location.  Thanks to the cryptographic properties of the hash, it is effectively impossible to generate a different identifier for the same content, or the same identifier for different content. We need not trust the authority of the registry, because we can always verify the content we recieve corresponds to what we wanted by comparing hashes (or hash URIs).
 
 
 ```{r}
-lookup("hash://sha256/9412325831dab22aeebdd674b6eb53ba6b7bdd04bb99a4dbb21ddff646287e37")
+query("hash://sha256/9412325831dab22aeebdd674b6eb53ba6b7bdd04bb99a4dbb21ddff646287e37")
 ```
 
 In this case, we see more than one URL has been registered containing exactly the same content. That is not surprising, as this dataset is commonly used and reproduced. Additional useful metadata, such as the file size and timestamp the URL was registered are also reported. Notably, multiple URLs for an object could serve different purposes: we could register URLs that are more suspetible to link rot in the long term but provide higher bandwidth for downloads in the short term (e.g. S3 bucket vs scientific repository.)
 
-It is worth noting that we still have no guarentee that these urls will not suffer link rot and cease to work -- this is not a replacement for archival storage efforts -- but the hash URI identifier allows us to avoid relying on a single storage point, and gives us a robust way to refer to an object.  
+It is worth noting that we still have no guarentee that these urls will not suffer link rot and cease to work -- this is not a replacement for archival storage efforts -- but the hash URI identifier allows us to avoid relying on a single storage point, and gives us a robust way to refer to an object. 
+
+## Local Registries & Local Storage
+
+We can also store and register content locally following this same scheme.
+
+```{r}
+store(co2_url)
+```
+
+Now when we query for the hash, we see a local storage location as well:
+
+```{r}
+query("hash://sha256/9412325831dab22aeebdd674b6eb53ba6b7bdd04bb99a4dbb21ddff646287e37")
+```
+
+
+The `retreive()` verb (subject to change), provides a wrapper around `query` that always returns exactly one local path to requested content.  It will check for a stored copy of the content first, and download (and verify) remote content otherwise:
+
+```{r}
+path <- retrieve("hash://sha256/9412325831dab22aeebdd674b6eb53ba6b7bdd04bb99a4dbb21ddff646287e37")
```

Comment:
I find it more intuitive to retrieve data from a store using the location that the store registered. It so happens to be that this local datastore also understands content hash uris. However, this might not be the case for data stores like Zenodo or the internet archive.

So, sequence would be:

```R
results <- registry::query("hash://sha256/ 941...")
# user picks suitable location
loc <- pick_favorite(results)
someData <- readLines(store::retrieve(loc))
```

with the pin() idea:

```R
# load registry / store state in the background
pin::load(someLocalFile)

readLines(pin("hash://sha256/9412325831dab22aeebdd674b6eb53ba6b7bdd04bb99a4dbb21ddff646287e37"))
# same result as 
readLines(pin("https://github.com/espm-157/climate-template/releases/download/data/vostok.icecore.co2"))
# because state of registry and store are such that the url is mapped to said content hash uri


# save registry / store state
pin::save(someLocalFile)

Reply:
Yes, retrieve definitely needs work!  looking forward to your comments on the questions in #15.

I'm not entirely sure what it means to retrieve by the location stored though.  If I do:
`store("https://example.com")` at two different times, I could store two different sets of content.  Which one is returned by `retrieve("https://example.com")` ?

Reply:
In my mind, you store data, not a location. But I see how you'd want to combine the retrieve and store verbs. 

How about ```retrieve_store_and_register("https://example.com", someRegistry)``` as a short cut for:
```R
url <- "https://example.com"
data <- curl(url)
(hash-uri,store_url) <- store(data)
register(hash-uri, store_url)
```
where store_url might or might not be accessible via https .

Reply:
the original url can then be submitted separately:

```
register(hash-uri, url)
```

Reply:
Conceptually, these two hash generation events are separate and linked. In practice, only a single hash is generated when saving the data in the local store and two registrations occurs.

Reply:
Ok, you lost me a few times over, we seem to be discussing a whole lot of different things here.  

I think your original question was whether it made sense to pass a content hash to `retrieve()`, or only a location. (currently it understands both, see #15, though the example above shows it using a content hash).  

I definitely want to be able to retrieve content by hash, ideally working across a distribution of registries.  But I can be convinced that retrieve should only take "a location" as input and not a hash.  What information is it supposed to return then?  This sounds like `query()` to me, so maybe we should make sure we agree about what `query()` does first (see #14).  In any event, you lost me in the bit about separate, linked, hash-generation events.  I see that come up with `store()` and `register()` potentially, but I thought we were discussing `retrieve()`, and as I've written it at least, neither `query()` nor `retrieve()` ever generate or register any hash.  (They are purely 'GET' / query methods, and don't change the state any registry or store)